packages/formik/src/useFormik.tsx

```diff
@@ -0,0 +1,512 @@
+  /** Manually set top level status. */
+  setStatus(status?: any): void;
+  /** Manually set errors object */
+  setErrors(errors: FormikErrors<Values>): void;
+  /** Manually set isSubmitting */
+  setSubmitting(isSubmitting: boolean): void;
+  /** Manually set touched object */
+  setTouched(touched: FormikTouched<Values>, shouldValidate?: boolean): void;
+  /** Manually set values object  */
+  setValues(values: Values, shouldValidate?: boolean): void;
+  /** Set value of form field directly */
+  setFieldValue(field: string, value: any, shouldValidate?: boolean): void;
+  /** Set error message of a form field directly */
+  setFieldError(field: string, message: string): void;
+  /** Set whether field has been touched directly */
+  setFieldTouched(
+    field: string,
+    isTouched?: boolean,
+    shouldValidate?: boolean
+  ): void;
+  /** Validate form values */
+  validateForm(values?: any): Promise<FormikErrors<Values>>;
+  /** Validate field value */
+  validateField(field: string): void;
+  /** Reset form */
+  resetForm(nextState?: Partial<FormikState<Values>>): void;
+  /** Submit the form imperatively */
+  submitForm(): Promise<void>;
+  /** Set Formik state, careful! */
+  setFormikState(
+    f:
+      | FormikState<Values>
+      | ((prevState: FormikState<Values>) => FormikState<Values>),
+    cb?: () => void
+  ): void;
+}
+
+export interface UseFormikOptions<Values> {
+  /**
+   * Initial values of the form
+   */
+  initialValues: Values;
+
+  /**
+   * Initial status
+   */
+  initialStatus?: any;
+
+  /** Initial object map of field names to specific error for that field */
+  initialErrors?: FormikErrors<Values>;
+
+  /** Initial object map of field names to whether the field has been touched */
+  initialTouched?: FormikTouched<Values>;
+
+  /**
+   * Submission handler
+   */
+  onSubmit: (
+    values: Values,
+    formikHelpers: FormikHelpers<Values>
+  ) => void | Promise<any>;
+  /**
+   * A Yup Schema or a function that returns a Yup schema
+   */
+  validationSchema?: any | (() => any);
+
+  /**
+   * Validation function. Must return an error object or promise that
+   * throws an error object where that object keys map to corresponding value.
+   */
+  validate?: (values: Values) => void | object | FormikErrors<Values>;
+}
+
+// Helper to map through field registry
+const forEachField = (
+  ref: React.MutableRefObject<FieldRegistry>,
+  cb: (key: string, field: FieldStateOpsAndRefs) => void
+) => {
+  Object.keys(ref.current).forEach(key => {
+    cb(key, ref.current[key]);
+  });
+};
+
+export function useFormik<Values extends FormikValues = FormikValues>({
+  // validateOnChange = true, // @todo
+  // validateOnBlur = true, // @todo
+  // validateOnMount = false, // @todo
+  // isInitialValid, // @todo
+  // enableReinitialize = false, // @todo
+  onSubmit,
+  validate,
+  initialValues,
+  validationSchema,
+}: UseFormikOptions<Values>) {
+  const fieldRegistry = React.useRef<FieldRegistry>({});
```

Comment:
Why not just use a `FieldRegistry` class here or `createFieldRegistry()` factory instead of using a dozen `useEventCallbacks()`? It can have a mutable `.fields` property and this way all of its methods will always operate on the latest `fields`. Wonder if I am missing something?

Reply:
`fieldRegistry` instance has to be preserved between renders and scoped to only this instance of `Formik`, hooks are currently the standard way to achieve that for a functional component.

It must be a ref in order to keep the values in sync across different hooks regardless of calling order.

This value will actually never change, so its inclusion in any hooks dependencies is just to satisfy eslint's `exhaustive-deps`. Use of `useEventCallback` is actually just so that those callbacks don't cause re-renders in child components (it's like useRef, but for callbacks).

Reply:
Exactly, the idea is to have a bunch of never changing references. Can't we just assign it like:

```js
const fieldRegistry = useRef(new FieldRegistry()).current

// Or even better, so as to not create a new registry on each render
const fieldRegistry = useMemo(() => new FieldRegistry(), [...])
```

The thinking is that all of the `useEventCallback()` calls can be dropped altogether. Furthermore if concurrent mode is considered to be an issue with this mutable source, this is just the right use case for the new `useMutableSource()` hook that @jaredpalmer mentioned above.